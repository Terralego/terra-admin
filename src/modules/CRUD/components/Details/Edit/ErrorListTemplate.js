import React from 'react';
import PropTypes from 'prop-types';
import { Callout } from '@blueprintjs/core';

const ErrorListTemplate = ({ errorSchema: { __errors, ...fields }, schema: { properties } }) => {
  const errors = Object.keys(fields).reduce((list, item) => {
    const { __errors: errorsItem } = fields[item];
    return errorsItem.length
      ? [...list, { title: properties[item].title, list: errorsItem }]
      : list;
  }, []);

  return (
    <Callout intent="danger">
      <ul>
        {errors.map(({ title, list }) => (
          <li key={title}>
            {title} :
            <ul>
              {list.map(item => <li key={item}>{item}</li>)}
            </ul>
          </li>
        ))}
      </ul>
    </Callout>
  );
};

ErrorListTemplate.propTypes = {
  errorSchema: PropTypes.shape({}).isRequired,
  schema: PropTypes.shape({
    properties: PropTypes.shape({}),
  }).isRequired,
};

export default ErrorListTemplate;
