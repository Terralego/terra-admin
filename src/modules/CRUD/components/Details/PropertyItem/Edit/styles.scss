@import "~@blueprintjs/core/src/components/forms/common";

.CRUD-edit {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-end;
  align-items: flex-end;
  &--vertical {
    flex-direction: column;
    align-items: stretch;
  }
  &--block > *:first-child {
    flex-basis: 100%;
  }
  fieldset {
    border: 0;
    margin: 0;
    padding: 0;
  }
  .error-list {
    margin: 0.5rem 1.5rem 1rem 0;
  }
  .control-label {
    display: block;
    margin-bottom: 0.2rem;
  }
  .form-group {
    margin-top: 0.4rem;
    margin-right: 0.2rem;
    flex-grow: 1;
    min-width: 0;
    &:first-child {
      margin-top: 0;
    }
    > * {
      width: 100%;
    }
  }
  .checkbox,
  .field-boolean {
    margin-top: 0.3rem;
    input {
      margin-right: 0.3rem;
    }
  }
  .radio-inline {
    margin-right: 1rem;
  }
  .field-radio-group {
    margin: 0.5rem 0;
  }
  input {
    &[type="text"],
    &[type="date"],
    &[type="number"] {
      @include pt-input();
    }
    &:disabled,
    &[readonly] {
      @include pt-input-disabled();
    }
    @include pt-input-placeholder();
  }
  select {
    @include pt-input();
    padding-right: 1rem;
    // Background below is the base 64 transformation of:
    // <svg viewBox="0 0 10 10">
    //   <polygon fill="#5c7080" points="5 0, 10 4, 0 4"/>
    //   <polygon fill="#5c7080" points="5 10, 10 6, 0 6"/>
    // </svg>
    background-image: url(data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgMTAgMTAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+ICA8cG9seWdvbiBmaWxsPSIjNWM3MDgwIiBwb2ludHM9IjUgMCwgMTAgNCwgMCA0Ii8+ICA8cG9seWdvbiBmaWxsPSIjNWM3MDgwIiBwb2ludHM9IjUgMTAsIDEwIDYsIDAgNiIvPjwvc3ZnPg==);
    background-position: calc(100% - 0.3rem) 50%;
    background-repeat: no-repeat;
    background-size: 0.6rem;
  }
  .help-block {
    opacity: 0.8;
  }
  .error-detail {
    padding: 0.2rem 0 0 1rem;
    color: red;
  }
  .field-array {
    legend {
      font-weight: bold;
      margin-bottom: 0.4rem;
    }
    .array-item,
    .array-item-add {
      display: flex;
      margin: 0.5rem 0 0;
      // Fields
      .col-xs-9 {
        flex-grow: 1;
      }
      // Controls
      .col-xs-3 {
        flex-shrink: 0;
      }
      .btn {
        @include pt-button-base();
        @include pt-button-height($pt-button-height);
        @include pt-button();
        margin-left: 0.2rem;

        &:disabled,
        &.#{$ns}-disabled {
          cursor: not-allowed;
        }
        &::before {
          margin: 0;
          color: #5c7080;
        }
        &.array-item-move-up::before {
          content: "↑";
        }
        &.array-item-move-down::before {
          content: "↓";
        }
        &.array-item-remove::before {
          content: "⨯";
          font-size: 1.5rem;
          margin-top: -0.5rem;
        }
        &.btn-add::before {
          content: "+";
          font-size: 1.1rem;
          margin-top: -0.3rem;
        }
        i {
          display: none;
        }
      }
    }
    .field-array-of-object .array-item-toolbox {
      padding-top: 1.3rem;
    }
  }
  .field-error .form-control {
    border: 1px solid #c23030;
  }
  &__submit {
    flex-shrink: 0;
    .details__list-create & {
      margin: 1rem 0.5rem 0 auto;
    }
  }
}
