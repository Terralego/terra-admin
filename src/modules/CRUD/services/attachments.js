import Api from '@terralego/core/modules/Api';
import { sanitizeCustomEndpoint } from './utils';

const attachmentRequest = (endpoint, body, method = 'PUT') =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/`, { method, body });

export const saveAttachment = attachmentRequest;

export const deleteAttachment = endpoint =>
  attachmentRequest(endpoint, {}, 'DELETE');

export const saveAttachmentCategories = body =>
  Api.request('crud/attachment-categories/', { method: 'POST', body });
