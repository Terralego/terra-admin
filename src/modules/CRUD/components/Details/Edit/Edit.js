import React from 'react';
import PropTypes from 'prop-types';
import Form from 'react-jsonschema-form';
import { Redirect } from 'react-router-dom';
import { Button } from '@blueprintjs/core';
import { CONTROL_DRAW, CONTROLS_TOP_LEFT } from '@terralego/core/modules/Map';
import RTEField from '../../../../../components/react-json-schemaForm/RTEField';

import { ACTION_CREATE, ACTION_UPDATE } from '../../../services/CRUD';
import { toast } from '../../../../../utils/toast';
import { generateURI } from '../../../config';
import Header from '../Header';
import Actions from '../Actions';
import ErrorListTemplate from './ErrorListTemplate';

import {
  ALL,
  POINT,
  LINESTRING,
  POLYGON,
  MULTI_POINT,
  MULTI_LINESTRING,
  MULTI_POLYGON,
} from '../../../../../utils/geom';

const updateSchemaPropertiesValues = (properties, formData) => (
  Object.keys(properties).reduce((list, prop) => ({
    ...list,
    [prop]: {
      ...properties[prop],
      ...(properties[prop].type === 'object')
        ? { properties: updateSchemaPropertiesValues(properties[prop].properties, formData[prop]) }
        : { default: formData[prop] },
    },
  }), {})
);
class Edit extends React.Component {
  static propTypes = {
    map: PropTypes.shape({}),
    feature: PropTypes.shape({
      title: PropTypes.string,
      properties: PropTypes.shape({}),
      geom: PropTypes.shape({}),
    }),
    saveFeature: PropTypes.func.isRequired,
    view: PropTypes.shape({
      formSchema: PropTypes.shape({}),
      layer: PropTypes.shape({ name: PropTypes.string }),
      name: PropTypes.string,
      uiSchema: PropTypes.shape({}),
    }),
    layerPaint: PropTypes.shape({}).isRequired,
    paramLayer: PropTypes.string.isRequired,
    paramId: PropTypes.string.isRequired,
    action: PropTypes.oneOf([ACTION_CREATE, ACTION_UPDATE]).isRequired,
    updateControls: PropTypes.func,
    getSettings: PropTypes.func,
    settingsEndpoint: PropTypes.string,
    displayAddFeature: PropTypes.bool,
    displayChangeFeature: PropTypes.bool,
    history: PropTypes.shape({
      push: PropTypes.func,
    }),
    t: PropTypes.func,
  }

  static defaultProps = {
    map: {},
    feature: {
      title: undefined,
      properties: {},
      geom: {},
    },
    view: {
      formSchema: {},
      layer: { name: undefined },
      name: undefined,
      uiSchema: {},
    },
    displayAddFeature: true,
    displayChangeFeature: true,
    updateControls () {},
    getSettings () {},
    settingsEndpoint: undefined,
    history: {
      push () {},
    },
    t: text => text,
  }

  state = {
    loading: false,
    schema: {
      properties: {},
    },
    geom: {},
  }

  componentDidMount () {
    this.setSchema();

    const {
      action,
      feature: { geom },
    } = this.props;

    if (action !== ACTION_UPDATE || Object.keys(geom).length) {
      this.initDraw();
    }
  }

  componentDidUpdate ({
    paramId: prevParamId,
    action: prevAction,
    feature: prevFeature,
    map: prevMap,
    view: prevView,
    featureError: prevFeatureError,
  }) {
    const {
      action,
      paramId,
      feature,
      map,
      view,
      featureError,
      t,
    } = this.props;
    if (prevFeature !== feature || prevMap !== map) {
      this.initDraw();
    }

    if (
      prevView !== view
      || prevParamId !== paramId
      || prevAction !== action
    ) {
      this.setSchema();
    }

    if (prevFeatureError !== featureError) {
      const { error: { data } } = featureError;
      toast.displayError(
        <div>
          {t(this.isActionUpdate ? 'CRUD.details.failUpdateFeature' : 'CRUD.details.failCreateFeature')}
          {Object.keys(data).map(item => <p key={item}>{item}: {data[item]}</p>)}
        </div>,
      );
    }
  }

  componentWillUnmount () {
    const {
      map,
      updateControls,
    } = this.props;
    // Remove controlDraw from controls
    updateControls([]);
    if (this.layerId && Object.keys(map).length) {
      map.setFilter(this.layerId, ['all']);
    }
  }

  get layerId () {
    const { layerPaint: { id } } = this.props;
    return id;
  }

  buildSchema = (schemaProperties, properties = {}) => {
    const {
      action,
    } = this.props;
    return Object.keys(schemaProperties).reduce((list, prop) => ({
      ...list,
      [prop]: {
        ...schemaProperties[prop],
        ...(schemaProperties[prop].type === 'object')
          ? { properties: this.buildSchema(schemaProperties[prop].properties, properties[prop]) }
          : {},
        ...(action !== ACTION_CREATE && properties[prop] && schemaProperties[prop].type !== 'object')
          ? { default: properties[prop] }
          : {},
      },
    }), {});
  }

  isActionUpdate = () => {
    const { action } = this.props;
    return action === ACTION_UPDATE;
  }

  setSchema = () => {
    const {
      feature: { properties = {}, geom = {} },
      view: { formSchema: schema = {} },
      t,
    } = this.props;
    if (!Object.keys(properties).length && !Object.keys(schema).length) {
      return;
    }
    this.setState(state => {
      const geometryFromMap = {
        type: 'boolean',
        title: t('CRUD.details.geometry'),
        default: !!Object.keys(state.geom).length,
      };
      return {
        geom,
        schema: {
          type: 'object',
          ...schema,
          properties: {
            ...this.buildSchema(schema.properties, properties),
            geometryFromMap,
          },
        },
      };
    });
  }

  initDraw = () => {
    const {
      map,
      paramId,
      feature: { geom },
      updateControls,
      action,
      view: { layer: { geom_type: geomType } },
    } = this.props;

    const control = {
      control: CONTROL_DRAW,
      position: CONTROLS_TOP_LEFT,
      onDrawUpdate: this.updateGeometry,
      onDrawCreate: this.updateGeometry,
      onDrawDelete: this.updateGeometry,
      controls: {
        point:  [ALL, POINT, MULTI_POINT].includes(geomType),
        line_string: [ALL, LINESTRING, MULTI_LINESTRING].includes(geomType),
        polygon:  [ALL, POLYGON, MULTI_POLYGON].includes(geomType),
        trash: true,
        combine_features: false,
        uncombine_features: false,
      },
    };

    if (action === ACTION_UPDATE) {
      if (!geom || !Object.keys(map).length) return;

      const listener = ({ control: addedControl }) => {
        if (addedControl !== control.control) {
          return;
        }
        map.draw.add(geom);
        map.off('control_added', listener);
      };
      map.on('control_added', listener);
      if (this.layerId) {
        map.setFilter(this.layerId, ['!=', '_id', paramId]);
      }
    }
    updateControls([control]);
  }

  updateGeometry = ({ type, features: [{ geometry: geom, id }], target: map }) => {
    if (type === 'draw.delete') {
      this.setState({ geom: {} });
      return;
    }

    const { view: { layer: { geom_type: geomType } } } = this.props;
    const { features } = map.draw.getAll();

    if (features.length > 1 && [POINT, LINESTRING, POLYGON].includes(geomType)) {
      map.draw.delete(
        features.reduce((list, feature) => (
          feature.id !== id
            ? [...list, feature.id]
            : list
        ), []),
      );
    }

    this.setState({ geom });
  }

  changeForm = ({ formData }) => {
    const { t } = this.props;

    this.setState(({ schema, geom }) => ({
      schema: {
        ...schema,
        properties: {
          ...updateSchemaPropertiesValues(schema.properties, formData),
          geometryFromMap: {
            type: 'boolean',
            title: t('CRUD.details.geometry'),
            default: !!Object.keys(geom).length,
          },
        },
      },
    }));
  }

  submitFeature = async ({ formData: { geometryFromMap, ...properties } }) => {
    const { geom } = this.state;
    const {
      getSettings,
      history: { push },
      view: { featureEndpoint },
      paramId,
      paramLayer,
      getFeaturesList,
      saveFeature,
      settingsEndpoint,
      t,
    } = this.props;

    this.setState({
      loading: true,
    });

    const savedFeature = await saveFeature(
      featureEndpoint,
      this.isActionUpdate ? paramId : false,
      { geom, properties },
    );

    if (savedFeature) {
      push(generateURI('layer', { layer: paramLayer, id: savedFeature.identifier }));
      getFeaturesList(
        featureEndpoint,
        { querystring: { page_size: 2000 } },
      );
      toast.displayToaster(
        { id: savedFeature.identifier },
        t(this.isActionUpdate ? 'CRUD.details.successUpdateFeature' : 'CRUD.details.successCreateFeature'),
      );
    } else {
      this.setState({
        loading: false,
      });
    }


    getSettings(settingsEndpoint);
  }

  validateForm = (formData, errors) => {
    const { geom } = this.state;
    const { t } = this.props;
    if (!Object.keys(geom).length) {
      errors.geometryFromMap.addError(t('CRUD.details.errorNoGeometry'));
    }
    return errors;
  }

  render () {
    const { loading, schema, schema: { properties } } = this.state;
    const {
      t,
      action,
      view: { layer: { name }, name: displayName = name, uiSchema = {} },
      paramLayer,
      paramId,
      displayAddFeature,
      displayChangeFeature,
      feature: { title },
    } = this.props;

    if (
      (action === ACTION_CREATE && !displayAddFeature)
      || (action === ACTION_UPDATE && !displayChangeFeature)
    ) {
      toast.displayError(t('CRUD.details.noAccess'));
      return (<Redirect to={generateURI('layer', { layer: paramLayer })} />);
    }

    const mainTitle = action === ACTION_CREATE
      ? t('CRUD.details.create', { layer: displayName })
      : (title || t('CRUD.details.noFeature'));

    const SaveButton = props => (
      <Button
        intent="primary"
        className="details__actions-save"
        loading={loading}
        {...props}
      >
        {action === ACTION_CREATE ? mainTitle : t('CRUD.details.save')}
      </Button>
    );

    const actionsButtons = {
      paramLayer,
      displayCancel: true,
      displayDelete: action === ACTION_UPDATE,
      ...action === ACTION_UPDATE ? { paramId } : {},
    };

    return (
      <div className="details ">
        <Header title={mainTitle} />
        <div className="details__content bootstrap-inside">
          {Object.keys(properties).length
            ? (
              <Form
                schema={schema}
                uiSchema={{ ...uiSchema, geometryFromMap: { 'ui:widget': 'hidden' } }}
                fields={{
                  rte: RTEField,
                }}
                onSubmit={this.submitFeature}
                onChange={this.changeForm}
                validate={this.validateForm}
                ErrorList={ErrorListTemplate}
              >
                <Actions {...actionsButtons}>
                  <SaveButton type="submit" />
                </Actions>
              </Form>
            )
            : (
              <Actions {...actionsButtons}>
                <SaveButton onClick={() => this.submitFeature({})} />
              </Actions>
            )
          }
        </div>
      </div>
    );
  }
}

export default Edit;
