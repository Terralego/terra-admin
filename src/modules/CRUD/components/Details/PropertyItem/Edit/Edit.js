import React, { useState, useEffect, useCallback, useMemo, useRef, useContext } from 'react';
import PropTypes from 'prop-types';
import { Button, Intent, Position, Tooltip } from '@blueprintjs/core';
import { useParams } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import Form from '@rjsf/core';
import classNames from 'classnames';
import { CRUDContext } from '../../../../services/CRUDProvider';
import { getView } from '../../../../services/CRUD';
import { requiredProperties } from '../../../../services/utils';
import customFields from '../../../../../../components/react-json-schemaForm';
import FileWidget from '../../../../../../components/react-json-schemaForm/FileWidget';
import ErrorListTemplate from '../../../../../../components/react-json-schemaForm/ErrorListTemplate';
import './styles.scss';

const widgets = {
  FileWidget: props => (
    <FileWidget {...props} />
  ),
};

const sanitizeValue = (schema, value) => {
  if (schema.type === 'array' && value === null) {
    return [];
  }
  return value;
};

const isFeatureToDelete = (isGeom, value, method) => {
  if (!isGeom || value.geom === null) {
    return false;
  }
  if (value.geom.coordinates.length === 0 && method === 'PATCH') {
    return true;
  }
  return false;
};

const Edit = ({
  editable,
  editedItem,
  isGeom,
  method,
  onEdit,
  onRead,
  name,
  schema,
  setEditedItem,
  settingsEndpoint,
  ui_schema: uiSchema,
  url,
  value: displayValue,
}) => {
  const { t } = useTranslation();
  const { id, layer } = useParams();

  const {
    feature,
    getFeaturesList,
    getSettings,
    settings,
    deleteFeature,
    saveFeature,
    fetchFeature,
  } = useContext(CRUDContext);

  const value = useMemo(() => {
    if (isGeom) {
      return sanitizeValue(schema, displayValue);
    }
    const { properties } = feature[id] || {};
    const propertiesListed = Object.values(properties).reduce(
      (list, item) => ({ ...list, ...item }
      ), {},
    );
    return sanitizeValue(schema, propertiesListed[name]);
  }, [displayValue, feature, id, isGeom, name, schema]);


  const {
    featureEndpoint,
    formSchema: { properties: schemaProperties } = {},
  } = useMemo(() => getView(settings, layer), [layer, settings]);

  const isMounted = useRef(true);

  const isCurrentEditedItem = editedItem === name;
  const canEdit =  ['', name].includes(editedItem) && editable;
  const isEdited = canEdit && isCurrentEditedItem;
  const [loading, setLoading] = useState(false);
  const [defaultValue, setDefaultValue] = useState(value);

  useEffect(() => {
    isCurrentEditedItem && !loading ? onEdit(name) : onRead(name);
  }, [isCurrentEditedItem, loading, name, onEdit, onRead]);

  useEffect(() => {
    isMounted.current = true;
    return () => { isMounted.current = false; };
  }, []);

  useEffect(() => {
    setDefaultValue(sanitizeValue(schema, value));
  }, [schema, value]);

  const handleSubmit = useCallback(async ({ formData }) => {
    const endpoint = url ? url.replace(featureEndpoint, '') : id;

    const [[propKey, propValue]] = Object.entries(formData);

    setLoading(true);

    const body = !isGeom
      ? { properties: { [propKey]: propValue === undefined ? '' : propValue } }
      : {
        geom: propValue.geom,
        ...(propValue.routingInformation && { routing_information: propValue.routingInformation }),
      };

    setDefaultValue(propValue);

    const isDeleting = isFeatureToDelete(isGeom, propValue, method);
    const saveOrDeleteFeature = isDeleting
      ? deleteFeature
      : saveFeature;

    const savedFeatures = await saveOrDeleteFeature(
      featureEndpoint,
      endpoint,
      body,
      method,
    );

    if (!isMounted.current) return;

    if (!savedFeatures.error) {
      getSettings(settingsEndpoint);
      if (isDeleting) {
        await fetchFeature(featureEndpoint, id);
      } else {
        getFeaturesList(featureEndpoint);
      }
      if (!isMounted.current) return;
      setEditedItem('');
      onRead(name);
    }
    setLoading(false);
  },
  [
    deleteFeature,
    featureEndpoint,
    fetchFeature,
    getFeaturesList,
    getSettings,
    id,
    isGeom,
    method,
    name,
    onRead,
    saveFeature,
    setEditedItem,
    settingsEndpoint,
    url,
  ]);

  const requiredField = useMemo(() => {
    const { required }  = requiredProperties(schemaProperties);
    return required.includes(name);
  }, [name, schemaProperties]);

  const action = useMemo(() => {
    if (isEdited) {
      return {
        label: t('CRUD.details.cancel'),
        icon: 'small-cross',
      };
    }
    if (!editable) {
      return {
        label:  t('CRUD.details.notAllowedToEdit'),
        icon:  'info-sign',
      };
    }
    return {
      label: t('CRUD.details.update'),
      icon: 'edit',
    };
  }, [editable, isEdited, t]);

  const enhanceUISchema = useMemo(() => {
    if (schema.type === 'string' && schema.enum?.length) {
      return { 'ui:placeholder': t('CRUD.details.placeholderSelect'), ...uiSchema };
    }
    return uiSchema;
  }, [schema.enum, schema.type, t, uiSchema]);

  return (
    <div className="details__list-edit">
      <Tooltip
        className="details__list-edit-button"
        content={action.label}
        position={Position.TOP}
        usePortal
        intent={!editable ? Intent.WARNING : Intent.INFO}
      >
        <Button
          icon={action.icon}
          disabled={!canEdit}
          onClick={() => setEditedItem(canEdit && !isCurrentEditedItem ? name : '')}
          minimal
        />
      </Tooltip>
      {isCurrentEditedItem && (
        <Form
          className={classNames({ 'CRUD-edit': true, 'CRUD-edit--block': isGeom })}
          disabled={loading}
          fields={customFields}
          ErrorList={ErrorListTemplate}
          onSubmit={handleSubmit}
          uiSchema={{ [name]: { ...enhanceUISchema } }}
          schema={{
            type: 'object',
            required: [requiredField && name].filter(Boolean),
            properties: {
              [name]: {
                ...schema,
                ...(defaultValue !== null && { default: defaultValue }),
              },
            },
          }}
          widgets={widgets}
        >
          {requiredField && (
            <span className="details__list-edit-mandatory details__list-edit-mandatory--edit">{t('CRUD.details.mandatory')}</span>
          )}
          <Button
            className="CRUD-edit__submit"
            loading={loading}
            type="submit"
          >
            {t('CRUD.details.save')}
          </Button>
        </Form>
      )}
    </div>
  );
};

Edit.propTypes = {
  editable: PropTypes.bool,
  editedItem: PropTypes.string,
  isGeom: PropTypes.bool,
  method: PropTypes.oneOf(['PATCH', 'POST', 'PUT']),
  onEdit: PropTypes.func,
  onRead: PropTypes.func,
  name: PropTypes.string,
  schema: PropTypes.shape({
    enum: PropTypes.arrayOf(PropTypes.string),
    properties: PropTypes.shape({}),
    type: PropTypes.string,
  }),
  setEditedItem: PropTypes.func,
  ui_schema: PropTypes.shape({}),
  url: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  value: PropTypes.any,
};

Edit.defaultProps = {
  editable: true,
  editedItem: '',
  isGeom: false,
  method: 'PATCH',
  onEdit: () => {},
  onRead: () => {},
  name: undefined,
  schema: {},
  setEditedItem: () => {},
  ui_schema: {},
  url: undefined,
  value: null,
};

export default Edit;
