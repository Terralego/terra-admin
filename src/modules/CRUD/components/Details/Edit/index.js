import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectAppProvider } from '../../../../../components/AppProvider';
import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { connectMapProvider } from '../../../views/Map/MapProvider';
import { withTableFilters } from '../../../services/UserSettingsProvider';
import { getView, getLayersPaints } from '../../../services/CRUD';
import compose from '../../../../../utils/compose';

import Edit from './Edit';

const appProviderGetter = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

const authProviderGetter = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayAddFeature: permissions.includes('can_add_feature'),
    displayChangeFeature: permissions.includes('can_change_feature'),
  };
};

const CRUDPRoviderGetter = ({
  getSettings,
  getFeaturesList,
  settings,
  map,
  feature,
  saveFeature,
  errors: {
    feature: featureError,
  },
}, {
  match: { params: { layer, id } },
}) => ({
  getSettings,
  getFeaturesList,
  map,
  feature: feature[id] || {},
  saveFeature,
  view: getView(settings, layer),
  layerPaint: getLayersPaints(settings).find(item => item['source-layer'] === layer) || {},
  paramLayer: layer,
  paramId: id,
  featureError: featureError.find(({ featureId }) => featureId === id),
});

export default compose(
  withRouter,
  connectAppProvider(appProviderGetter),
  connectAuthProvider(authProviderGetter),
  connectCRUDProvider(CRUDPRoviderGetter),
  connectMapProvider('addControl', 'removeControl'),
  withTableFilters(),
  withTranslation(),
)(Edit);
