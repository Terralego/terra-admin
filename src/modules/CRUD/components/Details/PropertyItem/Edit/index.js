import { connectAppProvider } from '../../../../../../components/AppProvider';

import Edit from './Edit';

import compose from '../../../../../../utils/compose';

const appProviderGetter = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

export default compose(
  connectAppProvider(appProviderGetter),
)(Edit);
