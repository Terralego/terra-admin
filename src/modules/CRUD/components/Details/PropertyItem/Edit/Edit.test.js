import React from 'react';
import renderer from 'react-test-renderer';

import Edit from './Edit';
import { CRUDContext } from '../../../../services/CRUDProvider';

jest.mock('react-router-dom', () => ({
  useParams: () => ({
    category: 'someCategory',
    id: 'layerId',
    layer: 'layerFoo',
    section: 'default',
  }),
}));

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: text => text,
  }),
}));

jest.mock('@blueprintjs/core', () => ({
  // eslint-disable-next-line react/button-has-type
  Button: props => <button {...props} />,
  Intent: {
    INFO: 'info',
    WARNING: 'warning',
  },
  Position: {
    TOP: 'top',
  },
  Tooltip: props => <div {...props} />,
}));
jest.mock('@rjsf/core', () => props => <form {...props} />);
jest.mock('../../../../../../components/react-json-schemaForm/FileWidget', () => props => <form {...props} />);
jest.mock('../../../../../../components/react-json-schemaForm/ErrorListTemplate', () => props => <div {...props} />);
jest.mock('../../../../../../components/react-json-schemaForm', () => ({
  GeometryField: props => <div {...props} />,
  RTEField: props => <div {...props} />,
  TableField: props => <div {...props} />,
}));

const props = {
  editedItem: '',
  method: 'PATCH',
  name: 'somePropertieName',
  setEditedItem:  text => text,
  schema: {
    somePropertieName: {
      foo: 'foo',
    },
  },
  ui_schema: {},
  value: 'null',
};

const providerProps = {
  feature: {
    layerId: {
      properties: [{}],
    },
  },
  getFeaturesList: () => {},
  getSettings: () => {},
  settings: {
    menu: [
      {
        crud_views: [
          {
            layer: {
              id: 8,
              name: 'layerFoo',
              settings: {
                tiles: {
                  properties_filter: [
                    'number',
                    'name',
                  ],
                },
              },
              geom_type: 0,
              routable: false,
            },
            form_schema: {
              required: [
                'name_fr',
              ],
              properties: {
                name_fr: {
                  type: 'string',
                  title: 'Nom FranÃ§ais',
                },
                type: {
                  enum: [
                    'Voie principale',
                    'Voie secondaire',
                    'Voie tertiaire',
                    'Voie alternative',
                    'Liaison',
                  ],
                  type: 'string',
                  title: 'Type',
                },
                name: {
                  type: 'string',
                  title: 'Nom',
                },
              },
            },
          },
        ],
      },
    ],
  },
  saveFeature: () => {},
};

describe('should render correctly', () => {
  it('can edit', () => {
    const tree = renderer.create((
      <CRUDContext.Provider value={providerProps}>
        <Edit
          {...props}
          editedItem="somePropertieName"
        />
      </CRUDContext.Provider>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('can\'t edit', () => {
    const tree = renderer.create((
      <CRUDContext.Provider value={providerProps}>
        <Edit
          {...props}
          editedItem="bar"
        />
      </CRUDContext.Provider>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('can\'t edit explicitly', () => {
    const tree = renderer.create((
      <CRUDContext.Provider value={providerProps}>
        <Edit
          {...props}
          editable={false}
          editedItem="bar"
        />
      </CRUDContext.Provider>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
