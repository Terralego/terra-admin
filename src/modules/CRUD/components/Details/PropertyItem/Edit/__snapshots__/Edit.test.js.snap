// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly can edit 1`] = `
<div
  className="details__list-edit"
>
  <div
    className="details__list-edit-button"
    content="CRUD.details.cancel"
    intent="info"
    position="top"
    usePortal={true}
  >
    <button
      disabled={false}
      icon="small-cross"
      minimal={true}
      onClick={[Function]}
    />
  </div>
  <form
    ErrorList={[Function]}
    className="CRUD-edit"
    disabled={false}
    fields={
      Object {
        "GeometryField": [Function],
        "RTEField": [Function],
        "TableField": [Function],
      }
    }
    onSubmit={[Function]}
    schema={
      Object {
        "properties": Object {
          "somePropertieName": Object {
            "default": undefined,
            "somePropertieName": Object {
              "foo": "foo",
            },
          },
        },
        "required": Array [],
        "type": "object",
      }
    }
    uiSchema={
      Object {
        "somePropertieName": Object {},
      }
    }
    widgets={
      Object {
        "FileWidget": [Function],
      }
    }
  >
    <button
      className="CRUD-edit__submit"
      loading={false}
      type="submit"
    >
      CRUD.details.save
    </button>
  </form>
</div>
`;

exports[`should render correctly can't edit 1`] = `
<div
  className="details__list-edit"
>
  <div
    className="details__list-edit-button"
    content="CRUD.details.update"
    intent="info"
    position="top"
    usePortal={true}
  >
    <button
      disabled={true}
      icon="edit"
      minimal={true}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`should render correctly can't edit explicitly 1`] = `
<div
  className="details__list-edit"
>
  <div
    className="details__list-edit-button"
    content="CRUD.details.notAllowedToEdit"
    intent="warning"
    position="top"
    usePortal={true}
  >
    <button
      disabled={true}
      icon="info-sign"
      minimal={true}
      onClick={[Function]}
    />
  </div>
</div>
`;
